src/pull-request-core.ts

```diff
@@ -0,0 +1,234 @@
+        sourceOwner,
+        sourceRepo,
+        destinationBranch,
+        sourceBranch,
+        subject,
+        description,
+        destinationOwner,
+        destinationRepo,
+        maintainerCanModify,
+        draft
+    }: PullRequest): Promise<Octokit.PullsCreateResponse> {
+        if (destinationOwner && destinationOwner !== sourceOwner) {
+            destinationBranch = `${sourceOwner}:${destinationBranch}`;
+        } else {
+            destinationOwner = sourceOwner;
+        }
+
+        if (!destinationRepo) {
+            destinationRepo = sourceRepo;
+        }
+
+        return Promise.resolve(
+            octokit.pulls
+                .create({
+                    owner: destinationOwner,
+                    repo: destinationRepo,
+                    title: subject,
+                    body: description,
+                    head: sourceBranch,
+                    base: destinationBranch,
+                    maintainer_can_modify: maintainerCanModify,
+                    draft
+                })
+                .then(({ data }) => data)
+        );
+    }
+
+    /**
+     * Pushes a commit of treeSha to commitBranch and returns the response.
+     */
+    function pushCommit(
+        owner: string,
+        repo: string,
+        commitBranch: string,
+        commitBranchSha: string,
+        treeSha: string,
+        authorName: string,
+        authorEmail: string,
+        commitMessage: string
+    ): Promise<GitUpdateRefResponse> {
+        return Promise.resolve(
+            octokit.git
+                .getCommit({
+                    owner,
+                    repo,
+                    commit_sha: commitBranchSha
+                })
+                .then(({ data }) =>
+                    octokit.git.createCommit({
+                        owner,
+                        repo,
+                        message: commitMessage,
+                        tree: treeSha,
+                        author: {
+                            name: authorName,
+                            email: authorEmail,
+                            date: new Date().toISOString()
+                        },
+                        parents: [data.sha]
+                    })
+                )
+                .then(({ data }) =>
+                    // update the branch ref to point to the created commit
+                    octokit.git.updateRef({
+                        owner,
+                        repo,
+                        ref: 'heads/' + commitBranch,
+                        sha: data.sha
+                    })
+                )
+                .then(({ data }) => data)
+        );
+    }
+
+    /**
+     * Creates a tree from the provided commitBranchSha ref for sourceFiles and returns the sha.
+     */
+    function createTree(
+        owner: string,
+        repo: string,
+        commitBranchSha: string,
+        sourceFiles: SourceFile[]
+    ): Promise<string[]> {
```

Comment:
This return type is not very telling. An improvement would be `Promise<[string, string]>` to declare that it is a tuple and documenting it represents the SHAs. Better would be `Promise<{ sha: string, commitBranchSha: string }>` to give semantic meaning to the two strings.